lib/msf/base/sessions/command_shell.rb

```diff
@@ -423,6 +426,40 @@ def repr(data)
     return data_repr
   end
 
+  def cmd_source_help
+    print_line("Usage: source [file] [background]")
+    print_line
+    print_line("Execute a local shell script file on remote machine")
+    print_line("This meta command will upload the script then execute it on the remote machine")
+    print_line
+    print_line("background")
+    print_line("`y` represent execute the script in background, `n` represent on foreground")
+  end
+
+  def cmd_source(*args)
+    if args.length != 2
+      # no argumnets, just print help message
+      return cmd_source_help
+    end
+
+    background = args[1].downcase == 'y'
+
+    local_file = args[0]
+    remote_file = "/tmp/." + ::Rex::Text.rand_text_alpha(32) + ".sh"
```

Comment:
You might want to protect the permissions of this file.

Reply:
Sorry, I didn't understand what you mean. The file created here is for temporary use and after use, it will be deleted. I don't understand why I need to protect the permissions of this file.

Reply:
@WangYihang: [TOCTOU](https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use) concerns about whether you're executing what you intended, or what someone with access to your system (or the path where this file is created on a remotely mounted FS) while you're working is intending (at time of use, of course).

Reply:
A good start is `0600` for a script executed directly by the shell.

Reply:
Okay,  I will update  the code.